text/3128-io-safety.md

```diff
@@ -0,0 +1,272 @@
+
+這些型別同時會實作既有的 `AsRawFd`、`IntoRawFd`，以及 `FromRawFd` 特徵，所以它們可和既有程式碼的 `RawFd` 型別交互使用。
+
+## `AsFd`、`Into<OwnedFd>` 與 `From<OwnedFd>`
+
+這些型別提供 `as_fd`、`into` 與 `from` 函式，類似於 `AsRawFd::as_raw_fd`、`IntoRawFd::into_raw_fd` 與 `FromRawFd::from_raw_fd`。
+
+## 原型實作
+
+上述所有原型放在：
+
+<https://github.com/sunfishcode/io-lifetimes>
+
+README.md 有文件鏈結、範例、和當前提供類似功能的 crate 之調查研究。
+
+# 缺點
+[缺點]: #缺點
+
+Crate 若用到檔案描述子，如 [`nix`] 或 [`mio`]，將需要遷移到有實作 `AsFd` 的型別，或將這類函式標示為不安全。
+
+crates 若用 `AsRawFd` 或 `IntoRawFd` 來接收任何「類檔案」或「類 socket」型別，如 [`socket2`] 的 [`SockRef::from`]，將需換成 `AsFd` 或 `Into<OwnedFd>`，或將這類函式標示為不安全。
+
+# 原理及替代方案
+[原理及替代方案]: #原理及替代方案
+
+## 有關「unsafe 僅為了記憶體安全性」
+
+Rust 有個慣例：`unsafe` 只適用於標示記憶體安全性。舉個有名的案例， [`std::mem::forget`] 曾標示為不安全，但後來[被改為安全]，且其結論指出 unsafe 僅該用作標示記憶體安全性，而不該用來標示可能作繭自縛的情況或恫嚇應避免使用的 API 上。
+
+記憶體安全性造成的危害比其他程式設計面向更甚，它不僅要避免意外行為發生，且仍需避免無法限制一段程式碼能做什麼。
+
+輸入輸出安全性也落在這個範疇，有兩個因素：
+
+- （若作業系統存在 `mmap` 相關 API）在安全封裝的 `mmap` 中，輸入輸出安全性的錯誤仍會導致記憶體安全性的錯誤。
+- 輸入輸出安全性之錯誤也意味著一段程式碼可以在沒有通知或被授予任何引用的情形下，讀寫或刪除程式其他部分正在使用的資料。這使得在無法通曉一個 crate 鏈結到的其他 crate 的所有實作細節下，非常難以限制這個 crate 可以做什麼。
+
+原始 handle 更像指向單獨的位址空間（address space）的原始指標，它們可能迷途（dangle）或造假。輸入輸出安全性近似於記憶體安全性，兩者皆竭力杜絕遠處來的詭異行為（spooky-action-at-a-distance），且對兩者來說，所有權都可作為建立穩健抽象化的主要根基，所以自然而然共用了相似的安全性概念。
+
+[`std::mem::forget`]: https://doc.rust-lang.org/std/mem/fn.forget.html
+[被改為安全]: https://rust-lang.github.io/rfcs/1066-safe-mem-forget.html
+
+## 將輸入輸出 Handle 當作純資料
+
+主要的替代方案的說法是原始 handle 為純資料（plain data），並沒有輸入輸出安全性，和作業系統資源的生命週期也無任何與生俱來的關係。至少在類 Unix 平台，這些永遠不會導致記憶體不安全或是未定義行為。
+
+不過，大部分的 Rust 程式碼不直接與原始 handle 互動。撇開本 RFC 不談，不與原始 handle 互動是件好事。所謂資源，一定會有生命週期，若大部分的 Rust 程式碼能使用各方面都更易上手又能自動管理生命週期的更高階的型別，這樣鐵定更棒。不過，純資料的方案對於相對不常見的案例，最多只能讓原始 handle 的操作容易撰寫些。這可能只是蠅頭小利，甚至可能是個缺點，有可能最後變相鼓勵大家在不需要時去用了原始 handle 。
+
+純資料的方案亦不需要變更任何 crate 的程式碼。而輸入輸出安全性的方案則需改動如 [`socket2`]、[`nix`] 和 [`mio`] 這些用到 [`AsRawFd`] 和 [`RawFd`] 的 crate，不過這改動可以漸進推廣到整個生態系，而不必一次性完成。
+
+## `IoSafe` 特徵（與它的前身 `OwnsRaw`）
+
+這個 RFC 在早先幾個版本提議過一個 `IoSafe` 特徵，這個特徵會帶來小程度但具侵入性的修復。來自該 RFC 的回饋促使一系列新型別與特徵的開發。這個開發牽扯更廣的 API 範圍，也意味著需要更多設計和審核。並且隨著時間推移，整個 crate 生態系需要更大規模的改動。然而，早期跡象指出，本 RFC 引入的新型別與特徵更易理解，使用上更順手且安全，所以長期來說有更穩健的基礎。
+
+`IoSafe` 早期叫做 `OwnsRaw`。我們很難替這個特徵找到恰到好處的名字，這也許是個訊號，表示它並非良好設計的特徵。
+
+# 先驅技術
+[先驅技術]: #先驅技術
+
+大部分記憶體安全的程式語言都對原始 handle 做了安全的抽象層。多數情況下，它們僅是簡單地避免暴露原始 handle，例如 [C#]、[Java] 等。若將透過原始 handle 執行輸入輸出標示為不安全，可讓 safe Rust 與這些程式語言達到相同程度的安全保證。
+
+在 crates.io 上有好幾個 crate 封裝了擁有或借用的檔案描述子。[io-lifetimes 的 README.md 中 Prior Art 一節]詳細描述了其與其他既有 crate 的同異。從高層次角度來看，既有的 crate 都與 io-lifetimes 共享相同的基本概念。這些 crate 都圍繞著 Rust 生命週期與所有權概念打造，而這恰恰說明這些概念非常適合這個問題。
+
+Android 有特殊的 API 會偵測不恰當的 `close`，詳見 rust-lang/rust#74860。這些 API 的動機就是一種輸入輸出安全性的應用。Android 的特殊 API 用了動態檢查，讓它們可以在跨來源語言的邊界實施這些規則。本 RFC 提議的輸入輸出安全性的型別和特徵只專注在 Rust 程式碼本身實施這些規則，所以它們可在編譯期間利用 Rust 的型別系統實施這些規則，而非延遲到執行期間。
```

Comment:
I know this was not linked in original RFC though, how about let's hyperlink it?

```suggestion
Android 有特殊的 API 會偵測不恰當的 `close`，詳見 [rust-lang/rust#74860](https://github.com/rust-lang/rust/pull/74680)。這些 API 的動機就是一種輸入輸出安全性的應用。Android 的特殊 API 用了動態檢查，讓它們可以在跨來源語言的邊界實施這些規則。本 RFC 提議的輸入輸出安全性的型別和特徵只專注在 Rust 程式碼本身實施這些規則，所以它們可在編譯期間利用 Rust 的型別系統實施這些規則，而非延遲到執行期間。
```